flink-core/src/main/java/org/apache/flink/api/common/typeutils/base/GenericArrayComparator.java

```diff
@@ -0,0 +1,205 @@
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public T[] readWithKeyDenormalization(T[] reuse, DataInputView source) throws IOException {
+		throw new UnsupportedOperationException();
+	}
+
+	@Override
+	public boolean invertNormalizedKey() {
+		return !ascendingComparison;
+	}
+
+	@Override
+	public int hash(T[] record) {
+		return Arrays.hashCode(record);
+	}
+
+	private int compareValues(Object first, Object second) {
+		if (first.getClass().equals(Boolean.class) && second.getClass().equals(Boolean.class)) {
+			return new BooleanComparator(true).compare((Boolean) first, (Boolean) second);
+		}
+		else if (first.getClass().equals(Byte.class) && second.getClass().equals(Byte.class)) {
+			return new ByteComparator(true).compare((Byte) first, (Byte) second);
+		}
+		else if (first.getClass().equals(Character.class) && second.getClass().equals(Character.class)) {
+			return new CharComparator(true).compare((Character) first, (Character) second);
+		}
+		else if (first.getClass().equals(Double.class) && second.getClass().equals(Double.class)) {
+			return new DoubleComparator(true).compare((Double) first, (Double) second);
+		}
+		else if (first.getClass().equals(Float.class) && second.getClass().equals(Float.class)) {
+			return new FloatComparator(true).compare((Float) first, (Float) second);
+		}
+		else if (first.getClass().equals(Integer.class) && second.getClass().equals(Integer.class)) {
+			return new IntComparator(true).compare((Integer) first, (Integer) second);
+		}
+		else if (first.getClass().equals(Long.class) && second.getClass().equals(Long.class)) {
+			return new LongComparator(true).compare((Long) first, (Long) second);
+		}
+		else if (first.getClass().equals(Short.class) && second.getClass().equals(Short.class)) {
+			return new ShortComparator(true).compare((Short) first, (Short) second);
+		}
+		else if (first.getClass().equals(String.class) && second.getClass().equals(String.class)) {
+			return new StringComparator(true).compare((String) first, (String) second);
+		}
+		return -1;
+	}
+
+	@SuppressWarnings("unchecked")
+	private int parseGenericArray(Object firstArray, Object secondArray) {
+		int compareResult = 0;
+		if (firstArray.getClass().isArray() && secondArray.getClass().isArray()) {
+			int min = Array.getLength(firstArray);
+			int tempResult = 0;
+
+			if (min < Array.getLength(secondArray)) {
+				tempResult = -1;
+			}
+			if (min > Array.getLength(secondArray)) {
+				min = Array.getLength(secondArray);
+				tempResult = 1;
+			}
+
+			for(int i=0; i < min; i++) {
+				int val = parseGenericArray(Array.get(firstArray, i), Array.get(secondArray, i));
+				if (val != 0 && compareResult == 0) {
+					compareResult = val;
+				}
+			}
+
+			if (compareResult == 0) {
+				compareResult = tempResult;
+			}
+		}
+		else {
+			compareResult = compareValues(firstArray, secondArray);
+		}
+		return compareResult;
+	}
+
+	@Override
+	public int compare(T[] first, T[] second) {
```

Comment:
You should check whether `componentInfo` is an `AtomicType` or a `ComponentType` and then create the comparators via the provided methods. You can take a look at the `ComponentType.createComparator` implementation.